### PR TITLE
fix build for 0.5.4

### PR DESCRIPTION
**Pull request by dtrifiro**
- deps: bump vllm-tgis-adapter to [0.3.0](https://github.com/opendatahub-io/vllm-tgis-adapter/releases/tag/0.3.0) to include 0.5.4 fixes
- deps: bump flashinfer to 0.1.2
- Dockerfile.ubi: force using python-installed cuda runtime libraries
- Dockerfile.ubi: downgrade build-stage CUDA to 12.1 for `torch==2.4.0` compatibility 
- Dockerfile: use uv pip everywhere (it's faster)

### CODE REVIEW

**Comment by dtrifiro** [comment]
This is an alternative approach to the fixes proposed in #126, which should keep the image size smaller because of the reuse of the cuda runtime libraries in the python venv

File: Dockerfile.ubi [hunk 1]
```diff
@@ -57,23 +57,30 @@ ENV PATH="$VIRTUAL_ENV/bin:$PATH"
...
+    --mount=type=cache,target=/root/.cache/uv \
     --mount=type=bind,source=requirements-common.txt,target=requirements-common.txt \
     --mount=type=bind,source=requirements-cuda.txt,target=requirements-cuda.txt \
-    pip install \
+    uv pip install \
         -r requirements-cuda.txt
 
+# force using the python venv's cuda runtime libraries
+ENV LD_LIBRARY_PATH="${VIRTUAL_ENV}/lib/python${PYTHON_VERSION}/site-packages/nvidia/cuda_nvrtc/lib:${LD_LIBRARY_PATH}"
+ENV LD_LIBRARY_PATH="${VIRTUAL_ENV}/lib/python${PYTHON_VERSION}/site-packages/nvidia/cuda_runtime/lib:${LD_LIBRARY_PATH}"
```
Comment (dtrifiro) [inline Dockerfile.ubi:68]: It looks like that even hardcoding `LD_LIBRARY_PATH` doesn't change how `cmake` finds nvrtc:

```
#25 10.18 -- CUDART: /usr/local/cuda/lib64/libcudart.so
#25 10.18 -- CUDA Driver: /usr/local/cuda/lib64/stubs/libcuda.so
#25 10.18 -- NVRTC: /usr/local/cuda/lib64/libnvrtc.so
```

but, assuming that the versions match (:facepalm:), the built libraries might be happy with what they find.,

File: Dockerfile.ubi [hunk 2]
```diff
@@ -151,22 +155,29 @@ WORKDIR /workspace
 ENV VIRTUAL_ENV=/opt/vllm
 ENV PATH=$VIRTUAL_ENV/bin/:$PATH
 
+# force using the python venv's cuda runtime libraries
+ENV LD_LIBRARY_PATH="${VIRTUAL_ENV}/lib/python${PYTHON_VERSION}/site-packages/nvidia/cuda_nvrtc/lib:${LD_LIBRARY_PATH}"
+ENV LD_LIBRARY_PATH="${VIRTUAL_ENV}/lib/python${PYTHON_VERSION}/site-packages/nvidia/cuda_runtime/lib:${LD_LIBRARY_PATH}"
+ENV LD_LIBRARY_PATH="${VIRTUAL_ENV}/lib/python${PYTHON_VERSION}/site-packages/nvidia/nvtx/lib:${LD_LIBRARY_PATH}"
```
Comment (dtrifiro) [inline Dockerfile.ubi:161]: The CUDA runtime libraries are installed as part of the standard `pip install torch`, but these are not the same that vLLM's extensions are compiled against, which system-installed (via `microdnf`) in the `cuda-base` docker stage.

In order to reduce the image size, we can skip installation via `microdnf` and set `LD_LIBRARY_PATH` to point to the libs in the python virtualenv to make sure that they are detected.

As mentioned in https://github.com/opendatahub-io/vllm/pull/52, we are linking the built vLLM extensions to the system's `libcudart.so.12`.  Before this patch this was marked as `not found` by `ldd`, but somehow not causing any issues up until now.

With 0.5.4 the built libraries started linking against `libnvrtc.so.12` and `libnvToolsExt.so.1`, this time causing `ImportError`s causing the server to fail to start:

```
WARNING 08-12 10:50:55 _core_ext.py:175] Failed to import from vllm._core_C with ImportError('libnvrtc.so.12: cannot open shared object file: No such file or directory')
Traceback (most recent call last):
  File "<frozen runpy>", line 189, in _run_module_as_main
  File "<frozen runpy>", line 112, in _get_module_details
  File "/opt/vllm/lib64/python3.11/site-packages/vllm/__init__.py", line 3, in <module>
    from vllm.engine.arg_utils import AsyncEngineArgs, EngineArgs
  File "/opt/vllm/lib64/python3.11/site-packages/vllm/engine/arg_utils.py", line 7, in <module>
    from vllm.config import (CacheConfig, DecodingConfig, DeviceConfig,
  File "/opt/vllm/lib64/python3.11/site-packages/vllm/config.py", line 11, in <module>
    from vllm.model_executor.layers.quantization import QUANTIZATION_METHODS
  File "/opt/vllm/lib64/python3.11/site-packages/vllm/model_executor/__init__.py", line 1, in <module>
    from vllm.model_executor.sampling_metadata import SamplingMetadata
  File "/opt/vllm/lib64/python3.11/site-packages/vllm/model_executor/sampling_metadata.py", line 11, in <module>
    from vllm.utils import (async_tensor_h2d, is_pin_memory_available,
  File "/opt/vllm/lib64/python3.11/site-packages/vllm/utils.py", line 29, in <module>
    from vllm import _custom_ops as ops
  File "/opt/vllm/lib64/python3.11/site-packages/vllm/_custom_ops.py", line 7, in <module>
    from vllm._core_ext import ScalarType
  File "/opt/vllm/lib64/python3.11/site-packages/vllm/_core_ext.py", line 177, in <module>
    ScalarType = torch.classes._core_C.ScalarType
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/vllm/lib64/python3.11/site-packages/torch/_classes.py", line 13, in __getattr__
    proxy = torch._C._get_custom_class_python_wrapper(self.name, attr)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Tried to instantiate class '_core_C.ScalarType', but it does not exist! Ensure that it is registered via torch::class_
```

Setting `LD_LIBRARY_PATH` to include these CUDA runtime libraries solves the linking issue, and seems to be working. 

We'll have to investigate if building with CUDA 12.4 and then linking at runtime to a different version might cause issues.

For extra context, `torch==2.4.0` requires `nvidia-cuda-runtime-cu12==12.1.105` and `nvidia-cuda-nvrtc-cu12==12.1.105`, so it probably makes sense to build vLLM with CUDA 12.1 (which I think is what upstream is using aswell).
```
torch==2.4.0
├── filelock [required: Any, installed: 3.15.4]
├── fsspec [required: Any, installed: 2024.6.1]
├── Jinja2 [required: Any, installed: 3.1.4]
│   └── MarkupSafe [required: >=2.0, installed: 2.1.5]
├── networkx [required: Any, installed: 3.3]
├── nvidia-cublas-cu12 [required: ==12.1.3.1, installed: 12.1.3.1]
├── nvidia-cuda-cupti-cu12 [required: ==12.1.105, installed: 12.1.105]
├── nvidia-cuda-nvrtc-cu12 [required: ==12.1.105, installed: 12.1.105]
├── nvidia-cuda-runtime-cu12 [required: ==12.1.105, installed: 12.1.105]
├── nvidia-cudnn-cu12 [required: ==9.1.0.70, installed: 9.1.0.70]
│   └── nvidia-cublas-cu12 [required: Any, installed: 12.1.3.1]
├── nvidia-cufft-cu12 [required: ==11.0.2.54, installed: 11.0.2.54]
├── nvidia-curand-cu12 [required: ==10.3.2.106, installed: 10.3.2.106]
├── nvidia-cusolver-cu12 [required: ==11.4.5.107, installed: 11.4.5.107]
│   ├── nvidia-cublas-cu12 [required: Any, installed: 12.1.3.1]
│   ├── nvidia-cusparse-cu12 [required: Any, installed: 12.1.0.106]
│   │   └── nvidia-nvjitlink-cu12 [required: Any, installed: 12.6.20]
│   └── nvidia-nvjitlink-cu12 [required: Any, installed: 12.6.20]
├── nvidia-cusparse-cu12 [required: ==12.1.0.106, installed: 12.1.0.106]
│   └── nvidia-nvjitlink-cu12 [required: Any, installed: 12.6.20]
├── nvidia-nccl-cu12 [required: ==2.20.5, installed: 2.20.5]
├── nvidia-nvtx-cu12 [required: ==12.1.105, installed: 12.1.105]
├── sympy [required: Any, installed: 1.13.2]
│   └── mpmath [required: >=1.1.0,<1.4, installed: 1.3.0]
├── triton [required: ==3.0.0, installed: 3.0.0]
│   └── filelock [required: Any, installed: 3.15.4]
└── typing_extensions [required: >=4.8.0, installed: 4.12.2
```

File: Dockerfile.ubi [hunk 3]
```diff
@@ -57,23 +57,26 @@ ENV PATH="$VIRTUAL_ENV/bin:$PATH"
 
 # install cuda and common dependencies
 RUN --mount=type=cache,target=/root/.cache/pip \
+    --mount=type=cache,target=/root/.cache/uv \
     --mount=type=bind,source=requirements-common.txt,target=requirements-common.txt \
     --mount=type=bind,source=requirements-cuda.txt,target=requirements-cuda.txt \
-    pip install \
+    uv pip install \
```
Comment (dtrifiro) [inline Dockerfile.ubi:63]: The `uv` stuff can be moved over to another PR if necessary, it was just helpful in speeding up the image build while debugging this.

**Comment by dtrifiro** [comment]
The smoke test has failed due to some issue with the cluster, but the image itself is working:

`quay.io/opendatahub/vllm:pr-127`

File: Dockerfile.ubi (hunk 2 above)
Comment (maxdebayser) [reply Dockerfile.ubi:162]: I can't see a revert in the 4 commits in this PR. 

Is the build working with the changes in this PR? In theory the changes make sense to me, but you never know until you build :cry:

Comment (prashantgupta24) [reply Dockerfile.ubi:162]: Daniele mentioned that the build did succeed? https://github.com/opendatahub-io/vllm/pull/127#issuecomment-2286856327

Comment (dtrifiro) [reply Dockerfile.ubi:162]: > I can't see a revert in the 4 commits in this PR.

I ended up just removing the commits altogether since it made little sense to have a change and then a revert.

Here's another PR based on this one using CUDA 12.1 https://github.com/opendatahub-io/vllm/pull/130, it should fail.

> Daniele mentioned that the build did succeed? https://github.com/opendatahub-io/vllm/pull/127#issuecomment-2286856327

yes, this build succeeds and the image works as expected (grpc and http server come up).

There's just an issue with env var parsing with `vllm-tgis-adapter==0.2.4`, I'll bump it as soon as a new version is released before merging this one.
